OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs

```diff
@@ -0,0 +1,177 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+using System.Collections.Generic;
+using System;
+using System.Drawing;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Deliver the unit in production via skylift.")]
+	public class ProductionAirdropInfo : ProductionInfo
+	{
+		public readonly string ReadyAudio = "Reinforce";
+
+		[Desc("Cargo aircraft used.")]
+		[ActorReference] public readonly string DeliveringActorType = "c17";
+
+		public override object Create(ActorInitializer init) { return new ProductionAirdrop(this, init.Self); }
+	}
+
+	class ProductionAirdrop : Production
+	{
+		public ProductionAirdrop(ProductionInfo info, Actor self)
+			: base(info, self) { }
+
+		public override Actor DoProduction(Actor self, ActorInfo actorToProduce, ExitInfo exitInfo, string raceVariant)
+		{
+			var exit = self.Location + exitInfo.ExitCell;
+			var spawn = self.CenterPosition + exitInfo.SpawnOffset;
+
+			var td = new TypeDictionary
+			{
+				new OwnerInit(self.Owner),
+				new LocationInit(exit),
+				new CenterPositionInit(spawn)
+			};
+
+			if (raceVariant != null)
+				td.Add(new RaceInit(raceVariant));
+
+			return self.World.CreateActor(false, actorToProduce.Name, td);
+		}
+
+		public override bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
+		{
+			var owner = self.Owner;
+			var actorInfos = actorsToProduce.ToList();
+
+			var info = (ProductionAirdropInfo)Info;
+			var deliveringActorType = info.DeliveringActorType;
+
+			// Start a fixed distance away: the width of the map.
+			// This makes the production timing independent of spawnpoint
+			var startPos = self.Location + new CVec(owner.World.Map.Bounds.Width, 0);
+			var endPos = new CPos(owner.World.Map.Bounds.Left - 5, self.Location.Y);
+
+			foreach (var tower in self.TraitsImplementing<INotifyDelivery>())
+				tower.IncomingDelivery(self);
+
+			owner.World.AddFrameEndTask(w =>
+			{
+				var altitude = self.World.Map.Rules.Actors[deliveringActorType].Traits.Get<PlaneInfo>().CruiseAltitude;
+				var a = w.CreateActor(deliveringActorType, new TypeDictionary
+				{
+					new CenterPositionInit(w.Map.CenterOfCell(startPos) + new WVec(WRange.Zero, WRange.Zero, altitude)),
+					new OwnerInit(owner),
+					new FacingInit(64)
+				});
+
+				Action func = null;
+				func = () =>
```

Comment:
This monster anonymous function should be split out into at least one proper method.